### PR TITLE
Day 04 in Standard ML

### PR DESCRIPTION
**Pull request by hypirion**

### CODE REVIEW

File: days/day-04/solutions/day04.sml [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+                [a, b] => (a, b)
+              | _ => raise Fail ("# of : <> 1 in " ^ s)
+    in
+    String.concatWith " " lines
+                      >> String.tokens (fn c => c = #" ")
+                      >> List.map split
+    end
+
+fun validInt numDigits low high s =
+    numDigits = String.size s andalso
```
Comment (Arxcis) [inline days/day-04/solutions/day04.sml:38]: andalso!! :clap: Sounds italian :stuck_out_tongue:

File: days/day-04/solutions/day04.sml [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+fun main () =
+    let val input = TextIO.inputAll TextIO.stdIn
+                                    >> String.fields (fn c => c = #"\n")
+                                    >> splitWith (fn c => c = "")
+                                    >> List.map passportFields
+    in
+        println $ Int.toString $ List.length $ List.filter hasRequiredFields input;
+        println $ Int.toString $ List.length $ List.filter validPassport input
+    end
+end
```
Comment (Arxcis) [inline days/day-04/solutions/day04.sml:80]: I like how this is the only place in the code which has side-effects. Everything else is pure functions right?

**Review by Arxcis — APPROVED** [review]
A lot of string-parsing today :+1: Happy to see you made it through :sweat_smile: Approved!

File: days/day-04/solutions/day04.sml (hunk 2 above)
Comment (hypirion) [reply days/day-04/solutions/day04.sml:80]: Yeah. `TextIO.inputAll TextIO.stdIn` and `println` are the only places where there's any IO/mutation.

**Pull request merged**